docs/proposals/210701_endpoint_discovery.md

```diff
@@ -0,0 +1,44 @@
+---
+title: Unified Endpoint discovery
+type: proposal
+menu: proposals
+status: XXX
+---
+
+### Related Tickets
+
+* https://github.com/thanos-io/thanos/issues/2600
+
+### Summary
+
+We want to propose a new flag called `--endpoint=<address>` that will be passed to Thanos query and replace the current `--store=<address>` and `--rule=<address>` flags, and add a discovery mechanism which will be used to determine what types of services this endpoint can serve.
+
+### Motivation
+
+Currently, in Thanos Query, the discovery of rules APIs happens via Store API's Info method, this makes it harder if we ever want to not have a coupling to the Store API (which is already planned for [scalable ruler proposal](https://github.com/thanos-io/thanos/blob/master/docs/proposals/202005_scalable-rule-storage.md)).
+
+We also require passing two different flags to the Thanos Query component `--store=<address>` and `--rule=<address>` and perform DNS discovery on often the same address multiple times, this can cause occasional DNS issues. Which is especially confusing when one works and the other doesn't.
```

Comment:
```suggestion
We also require passing two different flags to the Thanos Query component `--store=<address>` and `--rule=<address>` and perform DNS discovery on often the same address multiple times, **this can cause occasional DNS issues**. Which is especially confusing when one works and the other doesn't.
```
Maybe we can make this bold? Am I correctly understanding that this is basically the entire problem with the current approach? A sentence or two on what explicitly were the problems with DNS would be good to know/learn.

Reply:
The main problem is that we have to perform calls often twice because in a lot of use cases the endpoints we pass for ruler and store to the querier component, tend to be the same address. This for us at least caused DNS resolving issues, and one got resolved another did not. 

Another motivation is to also unite the discovery to be able to just pass an endpoint and not needing to couple the ruler to the store APIs as it is currently in the thanos querier, that is described in the initial issue and something we found as well when going through the code. So if just ruler is discovered we can at least potentially fetch rules and serve those. I believe this is also one of the uses cases for the work you are doing?

So I would say to sum up, both of those are the main motivation, not sure I would highlight one over the other, wdyt?